File: dashboard/backend/models/evaluation.go
// Package models defines data structures for the evaluation system.
package models

import (
	"time"
)

// EvaluationStatus represents the current state of an evaluation task.
type EvaluationStatus string

const (
	StatusPending   EvaluationStatus = "pending"
	StatusRunning   EvaluationStatus = "running"
	StatusCompleted EvaluationStatus = "completed"
	StatusFailed    EvaluationStatus = "failed"
	StatusCancelled EvaluationStatus = "cancelled"
)

// EvaluationDimension represents a type of evaluation to run.
type EvaluationDimension string

const (
	DimensionHallucination EvaluationDimension = "hallucination"
	DimensionReasoning     EvaluationDimension = "reasoning"
	DimensionAccuracy      EvaluationDimension = "accuracy"
	DimensionLatency       EvaluationDimension = "latency"
	DimensionCost          EvaluationDimension = "cost"
	DimensionSecurity      EvaluationDimension = "security"
)

// EvaluationConfig holds the configuration for an evaluation task.
type EvaluationConfig struct {
	Dimensions    []EvaluationDimension `json:"dimensions"`
	Datasets      map[string][]string   `json:"datasets"`        // dimension -> dataset names
	MaxSamples    int                   `json:"max_samples"`     // max samples per dataset
	Endpoint      string                `json:"endpoint"`        // router endpoint URL
	Model         string                `json:"model"`           // model to evaluate (empty for auto-discovery)
	Concurrent    int                   `json:"concurrent"`      // concurrent requests
	SamplesPerCat int                   `json:"samples_per_cat"` // samples per category (for reasoning)
}

// EvaluationTask represents an evaluation task stored in the database.
type EvaluationTask struct {
	ID              string           `json:"id"`
	Name            string           `json:"name"`
	Description     string           `json:"description"`
	Status          EvaluationStatus `json:"status"`
	CreatedAt       time.Time        `json:"created_at"`
	StartedAt       *time.Time       `json:"started_at,omitempty"`
	CompletedAt     *time.Time       `json:"completed_at,omitempty"`
	Config          EvaluationConfig `json:"config"`
	ErrorMessage    string           `json:"error_message,omitempty"`
	ProgressPercent int              `json:"progress_percent"`
	CurrentStep     string           `json:"current_step,omitempty"`
}

// EvaluationResult holds the results for a specific dimension/dataset combination.
type EvaluationResult struct {
	ID             string                 `json:"id"`
	TaskID         string                 `json:"task_id"`
	Dimension      EvaluationDimension    `json:"dimension"`
	DatasetName    string                 `json:"dataset_name"`
	Metrics        map[string]interface{} `json:"metrics"`
	RawResultsPath string                 `json:"raw_results_path,omitempty"`
}

// EvaluationHistoryEntry tracks metric values over time.
type EvaluationHistoryEntry struct {
	ID          int64     `json:"id"`
	ResultID    string    `json:"result_id"`
	MetricName  string    `json:"metric_name"`
	MetricValue float64   `json:"metric_value"`
	RecordedAt  time.Time `json:"recorded_at"`
}

// HallucinationMetrics holds parsed metrics from the hallucination benchmark.
type HallucinationMetrics struct {
	DatasetName                 string  `json:"dataset_name"`
	Endpoint                    string  `json:"endpoint"`
	Model                       string  `json:"model"`
	TotalSamples                int     `json:"total_samples"`
	SuccessfulRequests          int     `json:"successful_requests"`
	FailedRequests              int     `json:"failed_requests"`
	TotalHallucinationsDetected int     `json:"total_hallucinations_detected"`
	TruePositives               int     `json:"true_positives"`
	FalsePositives              int     `json:"false_positives"`
	TrueNegatives               int     `json:"true_negatives"`
	FalseNegatives              int     `json:"false_negatives"`
	Precision                   float64 `json:"precision"`
	Recall                      float64 `json:"recall"`
	F1Score                     float64 `json:"f1_score"`
	Accuracy                    float64 `json:"accuracy"`
	AvgLatencyMs                float64 `json:"avg_latency_ms"`
	P50LatencyMs                float64 `json:"p50_latency_ms"`
	P99LatencyMs                float64 `json:"p99_latency_ms"`
	FactCheckNeededCount        int     `json:"fact_check_needed_count"`
	DetectionSkippedCount       int     `json:"detection_skipped_count"`
	AvgContextLength            float64 `json:"avg_context_length"`
	EstimatedDetectionTimeMs    float64 `json:"estimated_detection_time_ms"`
	ActualDetectionTimeMs       float64 `json:"actual_detection_time_ms"`
	TimeSavedMs                 float64 `json:"time_saved_ms"`
	EfficiencyGainPercent       float64 `json:"efficiency_gain_percent"`
}

// ReasoningModeMetrics holds metrics for a single mode (standard or reasoning).
type ReasoningModeMetrics struct {
	ModeName             string  `json:"mode_name"`
	TotalQuestions       int     `json:"total_questions"`
	CorrectAnswers       int     `json:"correct_answers"`
	FailedQueries        int     `json:"failed_queries"`
	Accuracy             float64 `json:"accuracy"`
	AvgResponseTimeSec   float64 `json:"avg_response_time_sec"`
	AvgPromptTokens      float64 `json:"avg_prompt_tokens"`
	AvgCompletionTokens  float64 `json:"avg_completion_tokens"`
	AvgTotalTokens       float64 `json:"avg_total_tokens"`
	TokenUsageRatio      float64 `json:"token_usage_ratio"`
	TimePerOutputTokenMs float64 `json:"time_per_output_token_ms"`
}

// ReasoningImprovementSummary holds the comparison between modes.
type ReasoningImprovementSummary struct {
	AccuracyDelta             float64 `json:"accuracy_delta"`
	AccuracyImprovementPct    float64 `json:"accuracy_improvement_pct"`
	TokenUsageRatioDelta      float64 `json:"token_usage_ratio_delta"`
	TimePerOutputTokenDeltaMs float64 `json:"time_per_output_token_delta_ms"`
	ResponseTimeDeltaSec      float64 `json:"response_time_delta_sec"`
}

// ReasoningComparison holds the comparison results for a dataset.
type ReasoningComparison struct {
	Dataset            string                                     `json:"dataset"`
	Model              string                                     `json:"model"`
	Timestamp          string                                     `json:"timestamp"`
	StandardMode       ReasoningModeMetrics                       `json:"standard_mode"`
	ReasoningMode      ReasoningModeMetrics                       `json:"reasoning_mode"`
	ImprovementSummary ReasoningImprovementSummary                `json:"improvement_summary"`
	CategoryBreakdown  map[string]map[string]ReasoningModeMetrics `json:"category_breakdown,omitempty"`
}

// ReasoningMetrics holds the full output from the reasoning mode benchmark.
type ReasoningMetrics struct {
	GeneratedAt             string                `json:"generated_at"`
	Issue                   string                `json:"issue"`
	Title                   string                `json:"title"`
	Comparisons             []ReasoningComparison `json:"comparisons"`
	VSRConfigRecommendation interface{}           `json:"vsr_config_recommendation,omitempty"`
}

// DatasetInfo provides information about an available dataset.
type DatasetInfo struct {
	Name        string              `json:"name"`
	Description string              `json:"description"`
	Dimension   EvaluationDimension `json:"dimension"`
	SampleCount int                 `json:"sample_count,omitempty"`
}

// CreateTaskRequest is the request body for creating a new task.
type CreateTaskRequest struct {
	Name        string           `json:"name"`
	Description string           `json:"description"`
	Config      EvaluationConfig `json:"config"`
}

// RunTaskRequest is the request body for running a task.
type RunTaskRequest struct {
	TaskID string `json:"task_id"`
}

// ProgressUpdate represents a real-time progress update for SSE.
type ProgressUpdate struct {
	TaskID          string `json:"task_id"`
	ProgressPercent int    `json:"progress_percent"`
	CurrentStep     string `json:"current_step"`
	Message         string `json:"message,omitempty"`
	Timestamp       int64  `json:"timestamp"`
}

// ExportFormat specifies the output format for reports.
type ExportFormat string

const (
	ExportJSON ExportFormat = "json"
	ExportCSV  ExportFormat = "csv"
	ExportPDF  ExportFormat = "pdf"
)

File: dashboard/frontend/src/types/evaluation.ts
// Evaluation system TypeScript types

export type EvaluationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export type EvaluationDimension =
  | 'hallucination'
  | 'reasoning'
  | 'accuracy'
  | 'latency'
  | 'cost'
  | 'security';

export interface EvaluationConfig {
  dimensions: EvaluationDimension[];
  datasets: Record<string, string[]>; // dimension -> dataset names
  max_samples: number;
  endpoint: string;
  model: string;
  concurrent: number;
  samples_per_cat: number;
}

export interface EvaluationTask {
  id: string;
  name: string;
  description: string;
  status: EvaluationStatus;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  config: EvaluationConfig;
  error_message?: string;
  progress_percent: number;
  current_step?: string;
}

export interface EvaluationResult {
  id: string;
  task_id: string;
  dimension: EvaluationDimension;
  dataset_name: string;
  metrics: Record<string, unknown>;
  raw_results_path?: string;
}

export interface EvaluationHistoryEntry {
  id: number;
  result_id: string;
  metric_name: string;
  metric_value: number;
  recorded_at: string;
}

export interface DatasetInfo {
  name: string;
  description: string;
  dimension: EvaluationDimension;
  sample_count?: number;
}

export interface CreateTaskRequest {
  name: string;
  description: string;
  config: EvaluationConfig;
}

export interface RunTaskRequest {
  task_id: string;
}

export interface ProgressUpdate {
  task_id: string;
  progress_percent: number;
  current_step: string;
  message?: string;
  timestamp: number;
}

export interface TaskResults {
  task: EvaluationTask;
  results: EvaluationResult[];
}

// Dimension metadata for UI display
export const DIMENSION_INFO: Record<EvaluationDimension, { label: string; description: string; color: string }> = {
  hallucination: {
    label: 'Hallucination Detection',
    description: 'Measures the system ability to detect factual inaccuracies',
    color: '#ef4444', // red
  },
  reasoning: {
    label: 'Reasoning Mode',
    description: 'Compares standard vs reasoning mode performance',
    color: '#8b5cf6', // purple
  },
  accuracy: {
    label: 'Classification Accuracy',
    description: 'Measures routing and classification correctness',
    color: '#22c55e', // green
  },
  latency: {
    label: 'Latency',
    description: 'Measures response time and throughput',
    color: '#3b82f6', // blue
  },
  cost: {
    label: 'Cost Efficiency',
    description: 'Measures token usage and cost per request',
    color: '#f59e0b', // amber
  },
  security: {
    label: 'Security',
    description: 'Measures jailbreak and prompt injection detection',
    color: '#ec4899', // pink
  },
};

// Status metadata for UI display
export const STATUS_INFO: Record<EvaluationStatus, { label: string; color: string; bgColor: string }> = {
  pending: {
    label: 'Pending',
    color: '#6b7280',
    bgColor: 'rgba(107, 114, 128, 0.15)',
  },
  running: {
    label: 'Running',
    color: '#3b82f6',
    bgColor: 'rgba(59, 130, 246, 0.15)',
  },
  completed: {
    label: 'Completed',
    color: '#22c55e',
    bgColor: 'rgba(34, 197, 94, 0.15)',
  },
  failed: {
    label: 'Failed',
    color: '#ef4444',
    bgColor: 'rgba(239, 68, 68, 0.15)',
  },
  cancelled: {
    label: 'Cancelled',
    color: '#f59e0b',
    bgColor: 'rgba(245, 158, 11, 0.15)',
  },
};

// Helper functions
export function formatDuration(startTime?: string, endTime?: string): string {
  if (!startTime) return '-';
  const start = new Date(startTime).getTime();
  const end = endTime ? new Date(endTime).getTime() : Date.now();
  const durationMs = end - start;

  if (durationMs < 1000) return `${durationMs}ms`;
  if (durationMs < 60000) return `${(durationMs / 1000).toFixed(1)}s`;
  if (durationMs < 3600000) return `${Math.floor(durationMs / 60000)}m ${Math.floor((durationMs % 60000) / 1000)}s`;
  return `${Math.floor(durationMs / 3600000)}h ${Math.floor((durationMs % 3600000) / 60000)}m`;
}

export function formatDate(dateString?: string): string {
  if (!dateString) return '-';
  return new Date(dateString).toLocaleString();
}

export function getMetricValue(metrics: Record<string, unknown>, key: string): number | null {
  const value = metrics[key];
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const parsed = parseFloat(value);
    return isNaN(parsed) ? null : parsed;
  }
  return null;
}

export function formatMetricValue(value: number | null, format: 'percent' | 'decimal' | 'ms' | 'count' = 'decimal'): string {
  if (value === null) return '-';
  switch (format) {
    case 'percent':
      return `${(value * 100).toFixed(1)}%`;
    case 'ms':
      return `${value.toFixed(1)}ms`;
    case 'count':
      return value.toFixed(0);
    default:
      return value.toFixed(3);
  }
}
